### DISCUSSION TITLE
I want to extract drawings as images

### DISCUSSION THREAD

**Discussion by rafayaar** [pymupdf/PyMuPDF#3508] [Looking for help] [answered]
I am building a solution where when provided with PDF it can parse it and make it editable

I have extract text and images along with there bounding box's

What I want to achieve is, to actually extract drawings / shapes / graphics which is left on pdf and save them as png images
Of course, for this as well I would want bbox.

One of the approaches I tried is to use `.get_drawings()` function and map it on pixmap and save as Pixel file to image. But cant do it

Help required

**Comment by JorjMcKie** [comment, marked as answer]
If your vector graphics are made of drawing paths that are close enough to each other, the `Page` method [cluster_drawings](https://pymupdf.readthedocs.io/en/latest/page.html#Page.cluster_drawings) should deliver what you want.

Its parameters allow you some flexibility:
* You can preprocess the `get_drawings()` output to build some desired subset, which you can hand into the method. Otherwise, this parameter is also useful for avoiding multiple vector graphic extractions.
* You can control the "sensitivity" for regarding drawings to be "neighbours" by manipulating the tolerance values. The default is 3 points: two paths in `get_drawings()` are assumed to belong to the same cluster if their minimum distance is below 3.

**Reply by JorjMcKie** [reply]
I forgot to mention the obvious:
`cluster_drawings()` delivers a list of rectangles. You can take each of these bboxes and make an image from it, that can be saved:
```python
bboxes = page.cluster_drawings()
for i, bbox in enumerate(bboxes):
    pix = page.get_pixmap(clip=bbox)
    filename = f"{doc.name}-{page.number}-{i}.png"
    pix.save(filename)
```
So for the 5th vector graphics on page 3, a PNG image with the name `"input.pdf-3-5.png"` will be saved.

**Reply by JorjMcKie** [reply]
... ignoring 0-based numbers above 🤷‍♂️
